lib/websocket.cpp

```diff
@@ -0,0 +1,403 @@
+/// on_handshake
+/// @param ec the error code during connect
+/// @param
+void session::on_connect(beast::error_code ec,
+                         tcp::resolver::results_type::endpoint_type) {
+  if (ec) return fail(ec, "connect");
+
+  // Turn off the timeout on the tcp_stream, because
+  // the websocket stream has its own timeout system.
+  beast::get_lowest_layer(ws).expires_never();
+
+  // Set suggested timeout settings for the websocket
+  ws.set_option(
+      websocket::stream_base::timeout::suggested(beast::role_type::client));
+
+  // Set a decorator to change the User-Agent of the handshake
+  ws.set_option(
+      websocket::stream_base::decorator([](websocket::request_type &req) {
+        req.set(http::field::user_agent,
+                std::string(BOOST_BEAST_VERSION_STRING) +
+                    " websocket-client-async");
+      }));
+
+  // Perform the websocket handshake
+  ws.async_handshake(
+      host, "/",
+      beast::bind_front_handler(&session::on_handshake, shared_from_this()));
+}
+
+/// @brief Send the ping message once the handshake has been complete calls ->
+/// on_read
+/// @param ec error code on handshake
+void session::on_handshake(beast::error_code ec) {
+  if (ec) return fail(ec, "handshake");
+
+  // If you reach here connection is up set is_connected to true
+  is_connected = true;
+
+  // Start listening to the socket for messages
+  ws.async_read(
+      buffer, beast::bind_front_handler(&session::on_read, shared_from_this()));
+}
+
+/// @brief Deals with the message recieved from server, calls itself to keep
+/// listening to more messages
+/// @param ec error code while reading
+/// @param bytes_transferred Amount of byte recieved
+void session::on_read(beast::error_code ec, std::size_t bytes_transferred) {
+  boost::ignore_unused(bytes_transferred);
+  // report error on read
+  if (ec) {
+    return fail(ec, "read");
+  }
+
+  // id socket has been disconnected then return
+  if (!is_connected) {
+    return;
+  }
+
+  // the blocking can be seen by the fact that read is called once and not
+  // recursively
+  //std::cout << "Reading..." << std::endl;
+  // get the data from the websocket and parse it to json
+  auto res = buffer.data();
+  json data = json::parse(net::buffers_begin(res), net::buffers_end(res));
+  //std::cout << data << std::endl;
+  // denotes that this response is just after writing for subscription
+  if (justSubscribed) {
+    add_callback(data);
+  }
+
+  // denotes that this response is just after writing for unsubscription
+  else if (justUnsubscribed) {
+    remove_callback(data);
+  }
+
+  // this is a notification process it sccordingly
+  else {
+    call_callback(data);
+  }
+
+  // consume the buffer
+  buffer.consume(bytes_transferred);
+
+  // call read again to read continiously
+  ws.async_read(
+      buffer, beast::bind_front_handler(&session::on_read, shared_from_this()));
+}
+
+/// @brief function to write to the websocket
+void session::do_write() {
```

Comment:
Write logic is not async at all. You should avoid creating a seperate thread for write and only write on subscribe event and unsubscibe events. This will also get rid of queues.